### ISSUE TITLE
Feature Request: Add OnSelectionComplete() event

### ISSUE THREAD

**Issue by toddhd** [d3lm/ngx-drag-to-select#64]
Hi - 

I am using ngx-drag-select on a very large number of UI items (hundreds of elements). I need to process the items selected, but the process time takes a rather long time due to the number of objects involved. 

The hurdle I'm facing is that currently, I am using the onSelect() event to put the code that does the processing, but since that changes often and rapidly as you drag to select, this is slowing things down to an unacceptable speed. 

What would help a lot would be a simple way to know when the drag select is done, at which point, I could simply process the selected items once, instead of over and over a again. 

Is this something that can be added? Maybe a simple extra call on MouseUp()?

Thanks! (And happy holidays)

**Comment by d3lm** [comment]
Hey @toddhd, thanks again for using my library and opening this feature request. This is indeed a useful feature and we were already thinking about adding such outputs to the container. I'll work on this in the next days and this will soon be available.

@timdeschryver What outputs do you think would be useful?

Maybe `dtsSelectionStarted`, `dtsSelectionEnded`, anything else?

**Comment by d3lm** [comment]
There is now a PR #65 that introduces two new outputs - `selectionStarted` and `selectionEnded`. I think this should be enough for better and more efficient processing of the selected items.

@toddhd Do you think that this will do the trick in your case?

**Comment by toddhd** [comment]
> There is now a PR #65 that introduces two new outputs - `selectionStarted` and `selectionEnded`. I think this should be enough for better and more efficient processing of the selected items.
> 
> @toddhd Do you think that this will do the trick in your case?

Thank you, yes please, that will be perfect

**Comment by toddhd** [comment]
Yes and no. The event fires and that part "works", however it fires before
I'm done selecting. For example, if I hold down the shift key so that I can
select multiple objects in multiple locations, the event keeps firing in
between each selection even though I'm still holding down the shift key. My
expectation was that, when selecting multiple objects, that the
onSelectionEnded event would fire when I was actually done selecting
everything. Does that make sense?

**Comment by d3lm** [comment]
Uh! That must be a regression then. I thought I had tested this thoroughly. My bad then... I ll work on a patch. So sorry!

**Comment by d3lm** [comment]
But wait, what exactly do you mean? When I select one item and I hold down `shift` and select other items the `selectionEnded` event is only fired when I release the mouse button and not for every item that is emitted. Am I wrong?

**Comment by toddhd** [comment]
https://www.youtube.com/watch?v=-J7D6K3MNz4

See above.

If I understand what you just wrote, you said, "the selectionEnded event is
only fired when I release the mouse button" which is what is happening,
however what I expected was, if I'm still holding down the shift key, then
releasing the mouse button would NOT fire the event UNTIL I release the
shift key. Because until I release the shift key, I'm still not done adding
to the selection.

**Comment by d3lm** [comment]
Thanks for the video! But this is exactly what happens, no? Only that the `selectionEnded` event is not triggered when releasing your `shift` key but instead when releasing the left mouse button. All your selected items are propagated along with the event and you can access the list through `$event`. The thing is that the mouse button basically dictates when the selection starts and when it ends. The `shift` key is only used to switch to a different selection mode but you cannot start selecting items with only your keyboard. You always need to click and release the mouse button to end the selection. I am not really sure if it would make much sense to say that the event is only fired if both keyboard keys and mouse buttons are no longer pressed. But I might be wrong on this one.

**Comment by timdeschryver** [comment]
I tend to agree with @d3lm, I'm not sure that what you're asking is possible.
A user could also do the following: `shift + drag, release shift, shift + drag` - this would also trigger the event twice.

The first thing that popped in my mind, would be to also emit the "active" shortcuts together with the selected items. I'm not sure if this would make sense tho.

I also think that if you want to be performant, that adding a `debounce` or `throttle` to the `selectionEnded` would help. But this would be something you'd have to implement yourself into your app.

This is off topic, but I just wanted to say the following...
Thank you @toddhd for taking the time to create a detailed report 👏 !

**Comment by toddhd** [comment]
Okay. I will make this work :) Thank you both for such a fast and friendly
response! I will follow this and see what new features and fixes come up.
Plus, I always seem to learn more when watching how other coders solve a
problem.
